presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaMetadata.java

```diff
@@ -190,11 +212,103 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
     @Override
     public List<ConnectorTableLayoutResult> getTableLayouts(ConnectorSession session, ConnectorTableHandle table, Constraint<ColumnHandle> constraint, Optional<Set<ColumnHandle>> desiredColumns)
     {
-        InformationSchemaTableHandle handle = (InformationSchemaTableHandle) table;
-        ConnectorTableLayout layout = new ConnectorTableLayout(new InformationSchemaTableLayoutHandle(handle, constraint.getSummary()));
+        if (constraint.getSummary().isNone()) {
+            return ImmutableList.of();
+        }
+
+        InformationSchemaTableHandle handle = checkTableHandle(table);
+
+        Set<QualifiedTablePrefix> prefixes = calculatePrefixesWithSchemaName(session, constraint.getSummary(), constraint.predicate());
+        if (!handle.getSchemaTableName().equals(TABLE_SCHEMATA)) {
+            prefixes = calculatePrefixesWithTableName(session, prefixes, constraint.getSummary(), constraint.predicate());
+        }
+
+        ConnectorTableLayout layout = new ConnectorTableLayout(new InformationSchemaTableLayoutHandle(handle, prefixes));
         return ImmutableList.of(new ConnectorTableLayoutResult(layout, constraint.getSummary()));
     }
 
+    private Set<QualifiedTablePrefix> calculatePrefixesWithSchemaName(
+            ConnectorSession connectorSession,
+            TupleDomain<ColumnHandle> constraint,
+            Predicate<Map<ColumnHandle, NullableValue>> predicate)
+    {
+        Optional<Set<String>> schemas = stringFilter(constraint, SCHEMA_COLUMN_HANDLE);
+
+        if (schemas.isPresent()) {
+            return schemas.get().stream()
+                    .map(schema -> new QualifiedTablePrefix(catalogName, schema))
+                    .collect(toImmutableSet());
+        }
+
+        Session session = ((FullConnectorSession) connectorSession).getSession();
+        return metadata.listSchemaNames(session, catalogName).stream()
+                .filter(schema -> predicate.test(ImmutableMap.of(
+                        CATALOG_COLUMN_HANDLE, new NullableValue(createUnboundedVarcharType(), utf8Slice(catalogName)),
+                        SCHEMA_COLUMN_HANDLE, new NullableValue(createUnboundedVarcharType(), utf8Slice(schema)))))
+                .map(schema -> new QualifiedTablePrefix(catalogName, schema))
+                .collect(toImmutableSet());
+    }
+
+    public Set<QualifiedTablePrefix> calculatePrefixesWithTableName(
+            ConnectorSession connectorSession,
+            Set<QualifiedTablePrefix> prefixes,
+            TupleDomain<ColumnHandle> constraint,
+            Predicate<Map<ColumnHandle, NullableValue>> predicate)
+    {
+        Optional<Set<String>> tables = stringFilter(constraint, TABLE_COLUMN_HANDLE);
```

Comment:
Is this needed? Maybe we just apply predicate? Isn't this pre-optimization (maybe it isn't, idk)?